syntax = "proto2";

package NYT.NScheduler.NProto;

import "yt_proto/yt/core/misc/proto/guid.proto";
import "yt_proto/yt/core/misc/proto/error.proto";
import "yt_proto/yt/client/node_tracker_client/proto/node.proto";
import "yt_proto/yt/core/tracing/proto/tracing_ext.proto";
import "yt/ytlib/scheduler/proto/job.proto";
import "yt/ytlib/controller_agent/proto/controller_agent_service.proto";

////////////////////////////////////////////////////////////////////////////////

message TExecNodeDescriptor
{
    required int32 node_id = 1;
    required string address = 2;
    required double io_weight = 3;
    required bool online = 6;
    required TJobResources resource_limits = 4;
    required NNodeTrackerClient.NProto.TDiskResources disk_resources = 7;
    repeated string tags = 5;
}

message TOperationAlert
{
    required int32 type = 1; // EOperationAlertType
    required NYT.NProto.TError error = 2;
}

message TOperationAlerts
{
    repeated TOperationAlert alerts = 1;
}

message TOperationInfo
{
    optional NYT.NProto.TGuid operation_id = 1;
    repeated NYT.NControllerAgent.NProto.TTreeTaggedJobMetrics job_metrics = 2;
    optional TOperationAlerts alerts = 3;
    optional bytes suspicious_jobs = 4;
    optional NYT.NControllerAgent.NProto.TCompositeNeededResources composite_needed_resources = 6;
    repeated TJobResourcesWithQuota min_needed_resources = 7;
}

message TAgentToSchedulerOperationEvent
{
    required NYT.NProto.TGuid operation_id = 1;
    required int32 event_type = 2; // EAgentToSchedulerOperationEventType
    required int32 controller_epoch = 11 [default = -1];

    // for Suspended, Aborted, Failed
    optional NYT.NProto.TError error = 3;

    // for BannedInTentativeTree
    optional string tentative_tree_id = 4;

    // for BannedInTentativeTree
    repeated NYT.NProto.TGuid tentative_tree_job_ids = 5;

    // for InitializationFinished
    optional NYT.NControllerAgent.NProto.TInitializeOperationResult initialize_result = 6;

    // for PreparationFinished
    optional NYT.NControllerAgent.NProto.TPrepareOperationResult prepare_result = 7;

    // for MaterializationFinished
    optional NYT.NControllerAgent.NProto.TMaterializeOperationResult materialize_result = 8;

    // for RevivalFinished
    optional NYT.NControllerAgent.NProto.TReviveOperationResult revive_result = 9;

    // for CommitFinished
    optional NYT.NControllerAgent.NProto.TCommitOperationResult commit_result = 10;
}

message TAgentToSchedulerRunningJobStatistics
{
    required NYT.NProto.TGuid job_id = 1;

    required int64 preemptible_progress_time = 2;
}

message TSchedulerToAgentAbortedJobEvent
{
    required NYT.NProto.TGuid operation_id = 1;
    required NYT.NProto.TGuid job_id = 2;

    required NYT.NProto.TError error = 3;

    required bool scheduled = 4;

    optional uint64 finish_time = 5;
    optional int32 abort_reason = 6;
}

message TSchedulerToAgentOperationEvent
{
    required NYT.NProto.TGuid operation_id = 1;
    required int32 event_type = 2; // ESchedulerToAgentOperationEventType
}

message TScheduleJobSpec
{
    optional int64 waiting_job_timeout = 1;
}

message TScheduleJobRequest
{
    required NYT.NProto.TGuid operation_id = 1;
    required NYT.NProto.TGuid job_id = 2;
    required string tree_id = 3;
    required TJobResources job_resource_limits = 4;
    optional string pool_path = 8;
    required NNodeTrackerClient.NProto.TDiskResources node_disk_resources = 6;
    optional TScheduleJobSpec spec = 7;

    optional NYT.NTracing.NProto.TTracingExt tracing_ext = 9;
}

message TScheduleJobResponse
{
    message TFailedCounter
    {
        required int32 reason = 1; // EScheduleJobFailReason
        required int32 value = 2;
    }

    required NYT.NProto.TGuid job_id = 1;
    required bool success = 2;
    optional TJobResourcesWithQuota resource_limits = 3;
    optional bool interruptible = 4;
    repeated TFailedCounter failed = 5;
    required uint64 duration = 6;
    required NYT.NProto.TGuid operation_id = 7;
    required int32 controller_epoch = 8 [default = -1];

    optional NYT.NTracing.NProto.TTracingExt tracing_ext = 9;
}

message TEventQueueInbox
{
    required int64 next_expected_item_id = 1;
}

////////////////////////////////////////////////////////////////////////////////

message TReqHandshake
{
    required string agent_id = 1;
    required NYT.NNodeTrackerClient.NProto.TAddressMap agent_addresses = 2;
    repeated string tags = 3;
}

message TRspHandshake
{
    required NYT.NProto.TGuid incarnation_id = 1;
    required bytes config = 2; // YSON
    optional string scheduler_version = 3;
}

////////////////////////////////////////////////////////////////////////////////

message TReqHeartbeat
{
    message TAgentToSchedulerOperationEventsOutbox
    {
        required int64 first_item_id = 1;
        repeated TAgentToSchedulerOperationEvent items = 2;
    }

    message TAgentToSchedulerRunningJobStatisticsUpdatesOutbox
    {
        required int64 first_item_id = 1;
        repeated TAgentToSchedulerRunningJobStatistics items = 2;
    }

    required string agent_id = 1;
    required NYT.NProto.TGuid incarnation_id = 2;
    repeated TOperationInfo operations = 3;

    required bool exec_nodes_requested = 4;

    required TAgentToSchedulerOperationEventsOutbox agent_to_scheduler_operation_events = 7;
    required TAgentToSchedulerRunningJobStatisticsUpdatesOutbox agent_to_scheduler_running_job_statistics_updates = 15;

    required TEventQueueInbox scheduler_to_agent_aborted_job_events = 10;
    required TEventQueueInbox scheduler_to_agent_operation_events = 11;
    required TEventQueueInbox scheduler_to_agent_schedule_job_requests = 12;

    optional int64 controller_memory_limit = 13;
    optional int64 controller_memory_usage = 14;
}

message TRspHeartbeat
{
    message TSchedulerToAgentAbortedJobEventsOutbox
    {
        required int64 first_item_id = 1;
        repeated TSchedulerToAgentAbortedJobEvent items = 2;
    }

    message TSchedulerToAgentOperationEventsOutbox
    {
        required int64 first_item_id = 1;
        repeated TSchedulerToAgentOperationEvent items = 2;
    }

    message TExecNodeDescriptorList
    {
        repeated TExecNodeDescriptor exec_nodes = 1;
    }

    optional TExecNodeDescriptorList exec_nodes = 1;

    required TEventQueueInbox agent_to_scheduler_operation_events = 2;
    required TEventQueueInbox agent_to_scheduler_running_job_statistics_updates = 11;

    required TSchedulerToAgentAbortedJobEventsOutbox scheduler_to_agent_aborted_job_events = 4;
    required TSchedulerToAgentOperationEventsOutbox scheduler_to_agent_operation_events = 5;

    repeated NYT.NProto.TGuid operation_ids_to_unregister = 8;

    optional int32 operation_archive_version = 9;
    reserved 10;
}

////////////////////////////////////////////////////////////////////////////////

message TReqScheduleJobHeartbeat
{
    message TAgentToSchedulerScheduleJobResponsesOutbox
    {
        required int64 first_item_id = 1;
        repeated TScheduleJobResponse items = 2;
    }

    required string agent_id = 1;
    required NYT.NProto.TGuid incarnation_id = 2;
    required TAgentToSchedulerScheduleJobResponsesOutbox agent_to_scheduler_schedule_job_responses = 3;

    required TEventQueueInbox scheduler_to_agent_schedule_job_requests = 4;
}

message TRspScheduleJobHeartbeat
{
    message TSchedulerToAgentScheduleJobRequestsOutbox
    {
        required int64 first_item_id = 1;
        repeated TScheduleJobRequest items = 2;
    }

    required TEventQueueInbox agent_to_scheduler_schedule_job_responses = 1;

    required TSchedulerToAgentScheduleJobRequestsOutbox scheduler_to_agent_schedule_job_requests = 2;
}

////////////////////////////////////////////////////////////////////////////////
